### PR TITLE
8342651: Refactor array constant to use an array of jbyte

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This small patch refactors array constants in C2 to use an array of `jbyte`s instead of an array of `jvalue`. The former is much easier to work with and we can do `memcpy` with them trivially.

Since code buffers support alignment of the constant section, I have also allowed constant tables to be aligned more than 8 bytes and used it for constant vectors on machines not supporting `SSE3`. I also fixed an issue with code buffer relocation where the temporary buffer is not correctly aligned.

This patch is extracted from https://github.com/openjdk/jdk/pull/21229. Tests passed with `UseSSE=2` where 16-byte constants would be generated, as well as normal testing routines.

Please take a look and leave your reviews, thanks a lot.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8342651](https://bugs.openjdk.org/browse/JDK-8342651): Refactor array constant to use an array of jbyte (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21596/head:pull/21596` \
`$ git checkout pull/21596`

Update a local copy of the PR: \
`$ git checkout pull/21596` \
`$ git pull https://git.openjdk.org/jdk.git pull/21596/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21596`

View PR using the GUI difftool: \
`$ git pr show -t 21596`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21596.diff">https://git.openjdk.org/jdk/pull/21596.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21596#issuecomment-2424874941)
</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 1]
```diff
@@ -1700,25 +1700,36 @@ void C2_MacroAssembler::load_vector_mask(KRegister dst, XMMRegister src, XMMRegi
...
     case 4:  movdl(dst, src);   break;
     case 8:  movq(dst, src);    break;
     case 16: movdqu(dst, src);  break;
     case 32: vmovdqu(dst, src); break;
     case 64: evmovdqul(dst, src, Assembler::AVX_512bit); break;
     default: ShouldNotReachHere();
+    }
+  } else {
+    switch (vlen_in_bytes) {
+    case 4:  movflt(dst, src); break;
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:1715]: Indentation of the case statements needs to be fixed. Also above.

File: src/hotspot/cpu/x86/x86.ad [hunk 2]
```diff
@@ -2732,20 +2732,40 @@ void vec_spill_helper(C2_MacroAssembler *masm, bool is_load,
...
+static inline GrowableArray<jbyte>* vreplicate_imm(BasicType bt, T con, int len) {
+  int size = type2aelembytes(bt) * len;
+  GrowableArray<jbyte>* val = new GrowableArray<jbyte>(size, size, 0);
   for (int i = 0; i < len; i++) {
-    val->append(ele);
+    int offset = i * type2aelembytes(bt);
+    switch (bt) {
+    case T_BYTE: val->at(i) = con; break;
+    case T_SHORT: {
+      jshort c = con;
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/x86.ad:2743]: Why are these casts needed? Isn't `T con` already of the appropriate j-type?

**Review by TobiHartmann — COMMENTED** [review]
Looks good to me overall but reviewers who looked at [JDK-8278947](https://bugs.openjdk.org/browse/JDK-8278947) (@vnkozlov, @iwanowww) should have a look at this as well.

File: src/hotspot/cpu/x86/x86.ad (hunk 2 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad]: No for example when `bt == T_BYTE`, `T` is actual `jint`. As a result, I do this for all the cases for uniformity, also a mismatch will not result in a crash but may silently write the wrong data so I'm extra cautious here.

Comment (TobiHartmann) [reply src/hotspot/cpu/x86/x86.ad]: Makes sense, thanks for the clarification.

File: src/hotspot/share/opto/constantTable.cpp [hunk 3]
```diff
@@ -75,16 +62,45 @@ bool ConstantTable::Constant::operator==(const Constant& other) {
   }
 }
 
+int ConstantTable::alignment() const {
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/share/opto/constantTable.cpp:65]: I have added it to the header file.

File: src/hotspot/cpu/x86/x86.ad [hunk 4]
```diff
@@ -2762,20 +2762,40 @@ void vec_spill_helper(C2_MacroAssembler *masm, bool is_load,
...
-    default: ShouldNotReachHere();
-  }
+static inline GrowableArray<jbyte>* vreplicate_imm(BasicType bt, T con, int len) {
+  int size = type2aelembytes(bt) * len;
+  GrowableArray<jbyte>* val = new GrowableArray<jbyte>(size, size, 0);
   for (int i = 0; i < len; i++) {
-    val->append(ele);
+    int offset = i * type2aelembytes(bt);
+    switch (bt) {
+      case T_BYTE: val->at(i) = con; break;
```
[the remark this answers is not in the document]
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:2771]: `len` is at most 16 and is typically 1 (you only emit 1 element and the broadcast instruction will fill the whole register). Also, this function is only invoked a couple of times for each compilation and I think the compiler can do unswitching, too.

[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/cpu/x86/x86.ad:2771]: okay

**Review by vnkozlov — APPROVED** [review]
Good. Please push it into JDK 25 after 24 is forked.
We would need to re-test it internally before integration.

**Comment by merykitty** [comment]
@vnkozlov Thanks, I have merged the branch with the latest master

**Comment by vnkozlov** [comment]
Leonid is fixing issue with GHA testing.

**Review by vnkozlov — APPROVED** [review]
My tier1-4 testing passed.

**Comment by merykitty** [comment]
@TobiHartmann Could you please reapprove this PR, thanks very much.

**Review by TobiHartmann — APPROVED** [review]
Still good!

**Comment by merykitty** [comment]
Thanks very much for your reviews!
/integrate

**Pull request closed without merging**